### ISSUE TITLE
Evaluation for common Object Detection metrics

### ISSUE THREAD

**Issue by AtomScott** [AtomScott/SoccerTrack#16]
### Search before asking

- [X] I have searched the SoccerTrack [issues](https://github.com/AtomScott/SoccerTrack/issues) and found no similar feature requests.

### Description

An easier way to evaluate object detectors that I have trained

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by IkumaUchida** [comment]
相談事項。

- camera dataframeは、Ground truthと推論結果の値で同じ構造？
    - 例えば、推論時のconfidence scoreはどうする？

**Comment by AtomScott** [comment]
理想的には下記の形で実行できると良い。

```python
gt_df = ... # ファイルから読み込み
bbox_df = ... # 推論結果

from soccertrack.metrics import iou_score
iou_score(gt_df, bbox_df)
>>> 0.69 # 全フレームの平均
or
>>> [0.314, 0.42, ... ] # 平均せずにフレームごとで出力
```

上の例で`iou_score`の内部で統一的なフォーマット([`BoundingBox`](https://github.com/AtomScott/Python-Object-Detection-Metrics/blob/20ea52eba4935c75fe868186112b39abb6fb01dd/podm/bounding_box.py#L11)など)への変更を行うのは一つの手ではある。

```python
gt_bboxes = bbox_df.convert_to_bboxes()
gt_bboxes = gt_df.convert_to_bboxes()

type(gt_bboxes)
>>> list

type(gt_bboxes[0])
>>> soccertrack.utils.BoundingBox
```
といったメソッド（`convert_to_bboxes`）があれば、評価指標の実装にフォーカスしやすかったり、[object-detection-metrics](https://github.com/rafaelpadilla/Object-Detection-Metrics)も有効活用できる。

とはいえ、もし実装する指標が多くないならnumpy/scipyから実装した方がいい気もする（今後バグなどが出てきた時に対処しやすいため）。もちろん既存コードは参考にするべきだし、実装しやすい方でやればいいと思う。

調べてみると
* [IoU, AP, mAP](https://blog.deepblue-ts.co.jp/image-processing/objectdetection_ap/)
* [OCC](https://cyberagent.ai/blog/research/computervision/16366/)
が実装されていれば十分だと思うからそのつもりで実装方針を精緻化してもらっていい？

**Comment by AtomScott** [comment]
具体的には
1. 実装する各関数（iou_score, ap_score, map_score, occ_score）の入出力
2. BoundingBoxの仕様（引数・メソッドなど）。[Object-Detection-Metrics](https://github.com/rafaelpadilla/Object-Detection-Metrics)の[`BoundingBox`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBox.py), [`BoundingBoxes`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBoxes.py)で十分かどうか、必要・不要な関するがあるか、SoccerTrackに必要ない制約があるかなど
3. convert_to_bboxesの仕様&bbox_dataframeに追加する情報が必要かどうか

**Comment by IkumaUchida** [comment]
## 実装方針

### 実装の概要
（iou_score, ap_score, map_score）は[Object-Detection-Metrics](https://github.com/rafaelpadilla/Object-Detection-Metrics)の[`Evaluator`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/Evaluator.py)がカバーしていてそれを参考予定。

```python
gt_df = ... # ファイルから読み込み
bbox_df = ... # 推論結果

#Bbox dataframeをリストに変換。各要素にはBoundingBox形式のオブジェクトを格納
gt_bboxes = bbox_df.convert_to_bboxes() #or convert_to_bboxes()
dt_bboxes = gt_df.convert_to_bboxes()

type(gt_bboxes)
>>> list

type(gt_bboxes[0])
>>> soccertrack.utils.BoundingBox

#評価指標
metricsPerClass = evaluator.GetPascalVOCMetrics(boundingboxes, IOUThreshold=0.3)
print("Average precision values per class:\n")
# Loop through classes to obtain their metrics
ap_list = []
for mc in metricsPerClass:
    # Get metric values per each class
    c = mc['class']
    precision = mc['precision']
    recall = mc['recall']
    average_precision = mc['AP']
    ipre = mc['interpolated precision']
    irec = mc['interpolated recall']
    ap_list.append(average_precision)
    # Print AP per class
    print('%s: %f' % (c, average_precision))
```

>1.  実装する各関数（iou_score, ap_score, map_score, occ_score）の入出力

- iou_score ⇨ [`Evaluator().iou(boxA, boxB)`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/55038f0266eccc4944ef48360dbec4ea14de211d/lib/Evaluator.py#L385) を使用することで出力可能
    - 入力 : `boxA: list, boxB: list`(GTのBboxとdetectのBbox) -> float
    - 出力 : iou score
- ap_score ⇨ 上記コードのうち、`mc['AP']`に相当
    - 入力 : `mc: dict` -> float
    - 出力 : ap score
- map_score ⇨ 各クラスのap_score(`mc['AP']`)の平均を取ることで算出
    - 入力 : `ap_list: list` -> float
    - 出力 : ap score
 
- OCC ⇨  Object-Detection-Metricsには含まれていないが、[OC-costの著者がコードを公開している](https://github.com/mayu-ot/oc-cost)ので、`Evaluator`に組み込めると良さそう。
  

>2. BoundingBoxの仕様（引数・メソッドなど）。[Object-Detection-Metrics](https://github.com/rafaelpadilla/Object-Detection-Metrics)の[`BoundingBox`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBox.py), [`BoundingBoxes`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBoxes.py)で十分かどうか、必要・不要な関するがあるか、SoccerTrackに必要ない制約があるかなど

[`BoundingBox`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBox.py)と [`BoundingBoxes`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBoxes.py)の内容で十分であると思う。

>3. convert_to_bboxesの仕様&bbox_dataframeに追加する情報が必要かどうか

convert_to_bboxesの仕様について、下記の昨日を組み込めたら良いと考えている。
1. bbox_dataframeの各行の要素を[`BoundingBox`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/lib/BoundingBox.py)クラスの入力として、`bbox`のようなインスタンスを生成
2. 生成したインスタンスを`BoundingBoxes().addBoundingBox(bb)`](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/55038f0266eccc4944ef48360dbec4ea14de211d/lib/BoundingBoxes.py#L9)でappendして、リスト形式の`bboxes`を出力

`bbox_dataframe `に、`confidence score`の列を追加したい。`confidence score`はAP算出の際(precision&recall曲線を作る際)に使用する。

**Comment by IkumaUchida** [comment]
`gt_df`や`bbox_df `のタイプはpandas.core.frame.DataFrameだと思うけど、
```python
bbox_df.convert_to_bboxes()
```
の形にするなら、pandasにメソッドを追加する形になる気がするんだけどあってる？
その場合は[pandas_flavor](https://github.com/Zsailer/pandas_flavor)を使ってpandasにメソッドを追加する必要がありそう。

gt_bboxes =convert_to_bboxes(bbox_df)の形でbbox_dfを入力にした関数を作成する形にしたらpandas_flavorを使うことはなさそうと思った。

**Comment by AtomScott** [comment]
> gt_dfやbbox_df のタイプはpandas.core.frame.DataFrameだと思うけど、

Nope! [SoccerTrackMixin](https://github.com/AtomScott/SoccerTrack/blob/c7522106697c9d410c006b7673803bb17087de38/soccertrack/utils/DataFrame/base.py#L11)に追加する予定です。
